src/Symfony/Component/Validator/Constraints/DomainValidator.php

```diff
@@ -0,0 +1,75 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Symfony\Component\Validator\Exception\UnexpectedTypeException;
+use Symfony\Component\Validator\Exception\UnexpectedValueException;
+
+/**
+ * @author Dmitrii Poddubnyi <dpoddubny@gmail.com>
+ */
+class DomainValidator extends ConstraintValidator
+{
+    private const RESERVED = [
+        'example',
+        'invalid',
+        'localhost',
+        'test',
+    ];
+
+    public function validate($value, Constraint $constraint)
+    {
+        if (!$constraint instanceof Domain) {
+            throw new UnexpectedTypeException($constraint, __NAMESPACE__.'\Domain');
+        }
+
+        if (null === $value || '' === $value) {
+            return;
+        }
+
+        if (!is_scalar($value) && !(\is_object($value) && method_exists($value, '__toString'))) {
+            throw new UnexpectedValueException($value, 'string');
+        }
+
+        $value = (string) $value;
+        if ('' === $value) {
+            return;
+        }
+        if (!$this->isValid($value) || ($constraint->requireTld && !$this->isTld($value))) {
+            $this->context->buildViolation($constraint->message)
+                ->setParameter('{{ value }}', $this->formatValue($value))
+                ->setCode(Domain::INVALID_FORMAT_ERROR)
+                ->addViolation();
+        }
+    }
+
+    private function isValid(string $domain): bool
+    {
+        return false !== filter_var($domain, FILTER_VALIDATE_DOMAIN, FILTER_FLAG_HOSTNAME);
+    }
+
+    private function isTld(string $domain): bool
```

Comment:
I think we could improve this method name. `hasValidTld` seems a better name to me.

Reply:
`hasValidTld` would require to validate if the TLD is actually one of the millions of valid TLDs defined nowadays. I think it's "impossible" to validate that.

Reply:
but the current name is not really accurate and misleading. I think we could do better :) 
We could merge this method and the next one and rename it `hasReservedTld`.
There is no need to have 2 methods anyway

Reply:
IMHO The condition `$constraint->requireTld && !$this->hasValidTld($value)` better communicates the idea rather than if I merge `hasValidTld` into `hasReservedTld` and it'll be `$constraint->requireTld && $this->hasReservedTld($value)`

Reply:
>hasValidTld would require to validate if the TLD is actually one of the millions of valid TLDs defined nowadays. I think it's "impossible" to validate that.

@javiereguiluz we might check on the [Public Suffix List](https://publicsuffix.org/list/public_suffix_list.dat). Also there are [some](https://github.com/jeremykendall/php-domain-parser#usage) [libraries](https://github.com/layershifter/TLDDatabase#basic-usage).

Reply:
the Public Suffix List is not only about TLDs.